### ISSUE TITLE
add node properties to set channel, user name, topic

### ISSUE THREAD

**Issue by janvda** [travisghansen/node-red-slack#4]
I am just wondering if for the `slack web out` node and the `slack rtm out` we can not define some extra node properties that make it easier to use these nodes in the flows.

So I am thinking of the following things:

1. You can also specify the msg.topic in the node properties.  Moreover if the msg.topic is not set in the message or in the node properties then it will set it to the default value = "chat.postMessage" for slack web out and "message" for "slack rtm out".

2. You can also specify the slack channel in the node properties.  So if this field is specified and it is not specified in msg.payload.channel then it will copy the slack channel specified in node properties to msg.payload.channel.

3. You can also specify the username in the node properties.  So if this field is specified and it is not specified in msg.payload.username then it will copy the username specified in node properties to msg.payload.username

**Comment by janvda** [comment]
I would then also propose `slack web out` node that
1.  in case the msg.payload is a string that it is copied to the field msg.payload.text.
2.  in case the msg.payload is a number that it is converted into a string and copied to the field msg.payload.text.

**Comment by janvda** [comment]
For the `slack rtm in` I would propose  the following properties:
6. `channel` - this node will only listen to message posted in this channel.
7. `type` - specifying the specific type of events to listen for (e.g. "message") - it would be good if this is a drop down box only listing possible types.

**Comment by travisghansen** [comment]
@janvda I won't be adding any node properties that aren't applicable to everything the node provides.  ie: adding a `channel` property on the `slack-rtm-in` as it handles things that have nothing to do with channels.

I considered allowing for a `type` to be selected on the `slack-rtm-in` node but ultimately decided against because at least:
1. the list of events is extensive
1. if the list changes in the slack API I didn't want to have to update the package
1. the API allows for subscribing to `subtype`s which would add even more headaches to the UI

It's quite easy to add a `switch` node immediately after the `slack-rtm-in` node and redirect based on `type` property.

I could still be convinced to add `type` to node properties though.

Regarding points 4 and 5: if the method (either explicitly set by using the proper name or using the special `@` or `#` syntax) is for sending a message, the payload (whether `msg.payload` or `msg.payload.text`) is forced to be a string already.  In essence, if the node knows it's sending a message it behaves nicely on the user's behalf.

In short, most of what you've asked for is easily accomplished with simple `function` or `switch` nodes before/after the appropriate `slack-foo` node.  In an effort to keep the code clean and concise I'm generally shying away from 'magic' handling/use-cases.  The only exception (which I nearly didn't add) is the existing `@` and `#` `topic` syntax.

**Comment by janvda** [comment]
I think you mainly reacted to my last comment (points 6 and 7) and not to my other points 1,2,..5.

I understand that you can do everything with switch or function nodes, but I still think that it would be good that you don't need these extra nodes for the basic use cases.  I think it is also good to be able to specify the things that are static in the node properties and the things that are dynamic through the msg properties passed to it.  It makes your flows easier to understand.  Take also care that a substantial subset of the users of node-red are not fluent javascript programmers (like you clearly are).

I know you can use @ and # in the msg.topic but in my use case I need to set both user name and channel so I can not resort to this.  I also admit that I am not really fond of this (I would rather prefer this to be set in the node properties).

In case of the `type` it is fine for me if this is a string instead of a drop down list.
But also in case this is not specified then use the best default type.

**Comment by travisghansen** [comment]
@janvda I addressed 1-5 in my first comment about not adding node properties that aren't applicable to everything the node provides.

It sounds like if I did the following you'd be covered for your wish list:

1. Made a new node designed specifically for sending messages that optionally allowed setting all properties on the node
1. Allowed for setting desired event types on the `slack-rtm-in` node

Is that a fair assessment?

**Comment by janvda** [comment]
@travisghansen thanks for the response.

Regarding 1.   Yes I agree with that.  I am only wondering if we really need a new node for that and if that can not be integrated in the current `slack-web-out` node.

Regarding 2.  Yes I agree with that.

**Comment by travisghansen** [comment]
@janvda yeah, and having the combination of both would leave you with nothing else on the wish list yeah?

Let me put some some thought into each of these and see where things land.

**Comment by janvda** [comment]
@travisghansen note that it is not really about me or my wish list.  It is more about the fact that `node-red-contrib-slack` according flows.nodered.org is the 11th most popular node in terms of downloads.  So there are certainly several other users already using these nodes.  I was more thinking about them.

**Comment by travisghansen** [comment]
@janvda fair enough..

**Comment by travisghansen** [comment]
@janvda I'm still not sure what I'm going to do here FYI.  I'm torn.  No matter what I'll be adding a exhaustive example of sending a message with all attributes using a `function` node with `slack-web-out`.

**Comment by janvda** [comment]
I think it is important to focus on those things that the end user most likely would like to specify as a node property.  So it is not important to be exhaustive.

So for the node `slack web out` I think this is the topic, channel and user.  Maybe you can also consider the node properties that are currently supported by the existing `slack` node.

So for node `slack rtm in` this is a bit less important, I think that `type` is a good candidate.

**Comment by travisghansen** [comment]
Yeah, the `slack-web-out` node handles stuff where those options are meaningless though.  That's more confusing (IMO) than not having the options at all.

I'm prototyping the `slack-rtm-in` option to select which events they want to subscribe to but it's not as straight forward as it seems.  We'll see where it goes..

**Comment by travisghansen** [comment]
@janvda I've decided to not introduce a new node but I've added a simple example for sending messages with the `slack-web-out` node.

I have also added a feature allowing to set a node property to filter events on the `slack-rtm-in` nodes.  Review the documentation in the `README.md` or in the palette and let me know your thoughts.

**Comment by janvda** [comment]
I accept your decision.  I consider this issue closed !! Thanks a lot for all the great effort !
Jan.
